README.md
# Walts：一款高可用、支持多种大模型、功能丰富的强大的AI编程助手

### Overview

Walts(读音同Waltz, 这既是我们四个开发者名字首字母的缩写，也代表着我们祝愿所有使用本插件的开发者能够像华尔兹一样优雅编程)是一款高可用（前后端分离，后端部署在服务器上，提供稳定的服务）、支持多种大模型（openai系列、智谱ai的Chatglm、星火认知大模型等）、功能丰富（提供诸如代码解释、代码优化、代码检查等功能）的强大的AI编程助手。


### 需求分析

#### 功能性需求

- 用户可以通过对话框向大模型询问代码问题
- 用户可以选择一段代码，然后要求给出代码的解释
- 用户可以选择一段代码，然后要求优化这段代码
- 用户可以选择一段代码，然后要求检查这段代码中的潜在问题，并提供修改意见
- 用户可以选择一段代码，然后要求在不改变代码功能的前提下重构这段代码
- 用户可以选择一段代码，然后要求给出描述这段代码的文档
- 用户可以自主选择一些参数，如model, temprature, top_p等

#### 非功能性需求
- 系统应当提供稳定的服务，不能出现因为超时访问服务器而导致的服务不可以现象
- 由于使用本插件的未必母语是英语，所以系统应该考虑到这一点提供高质量的服务
- 系统的输出应当方便用户，比如说将代码部分显示出来

### UI Design

### 接口设计

本插件设计为前后端分离，其中前端用typescript、css、html、javascript等语言构建，主要负责呈现界面，后端用python实现，负责具体功能的实现，部署在服务器上（由于缺乏云服务器，且大模型的apiKey均为个人账号，部署在云上供任意人使用容易出现安全隐患，所以暂时部署在实验室的服务器上，校园网内的用户均可以正常访问，正常使用插件提供的高质量服务）。后端只提供一个接口，如下：
接口地址：`/api/search`
请求方式：`POST`
请求参数：
 - `vendor`: 服务提供商，如openai,zhipuai,sparkai等
 - `model`: 具体模型，各个服务提供商的可选模型是不同的，如openai可选模型有text-davinci-003、text-davinci-002等
 - `APIKey`: 对于所有服务提供商，都提供免费的公共apiKey, 但用户也可以传入自己的apiKey,系统保证只使用该key调用相应的大模型接口，不会存储滥用用户传入的APIKey
 - `promptType`: 提示的类型，可选有"plain"(表示用户简单的询问)， "code explain"(用户选择一段代码要求对其进行解释，该值不为plain时，prompt应为空)
 - `prompt`: 用户输入的提示，仅当`promptType`为plain时才有效
 - `code`: 用户选中的代码片段，仅当`promptType`不为plain时才有效
 - `maxTokens`、`temprature`、`topP`为调用大模型接口使用的参数
 - `enableTranslate`: boolean类型，仅当调用英文大模型时有效，如果为true,将用户的prompt翻译成英文
响应参数：response, 即位大模型输出的文本


### 功能实现细节

### 使用手册

#### 使用前提
从插件市场

### 团队分工



CHANGELOG.md
# Change Log

All notable changes to the "walts" extension will be documented in this file.

Check [Keep a Changelog](http://keepachangelog.com/) for recommendations on how to structure this file.

## [Unreleased]

- Initial release

src/util.py
# -*- coding: utf-8 -*-
import sys
import uuid
import requests
import hashlib
import time
import json


def encrypt(signStr):
    hash_algorithm = hashlib.sha256()
    hash_algorithm.update(signStr.encode('utf-8'))
    return hash_algorithm.hexdigest()


def truncate(q):
    if q is None:
        return None
    size = len(q)
    return q if size <= 20 else q[0:10] + str(size) + q[size - 10:size]


def do_request(data, url):
    headers = {'Content-Type': 'application/x-www-form-urlencoded'}
    return requests.post(url, data=data, headers=headers)

        
def translate(text: str, youdaoConfig: dict) -> str:
    appid, app_secret, url = youdaoConfig['appid'], youdaoConfig['app_secret'], youdaoConfig['url']
    data = {}
    data['from'] = 'auto'
    data['to'] = 'en'
    data['signType'] = 'v3'
    curtime = str(int(time.time()))
    data['curtime'] = curtime
    salt = str(uuid.uuid1())
    signStr = appid + truncate(text) + salt + curtime + app_secret
    sign = encrypt(signStr)
    data['appKey'] = appid
    data['q'] = text
    data['salt'] = salt
    data['sign'] = sign
    #data['domain'] = 'computers'
    response = do_request(data, url)
    res = response.content.decode('utf-8')
    res = json.loads(res)
    return res['translation'][0]


if __name__ == '__main__':
    print(translate('你好', {
        "appid": "2c35f8f88f85ee4c",
        "app_secret": "0FniYwQvIcgB7Zihe7SOwYxKHLkAPUIY",
        "url": "https://openapi.youdao.com/api"
    }))

src/extension.ts
import * as vscode from 'vscode';
import axios from 'axios'


type KeyInfo = {apiKey?: string};
export type Settings = {pasteOnClick?: boolean, model?: string, maxTokens?: number, temperature?: number, enableTranslate?: boolean, vendor?: string, topP?: number};

export function activate(context: vscode.ExtensionContext) {
	
	// 创建页面视图，将其注册到vscode的上下文中
	const provider = new waltsViewProvider(context.extensionUri);
	
	// 获取用户设置
	const config = vscode.workspace.getConfiguration('walts');
	// 将用户设置的内容传递给页面视图
	provider.setAPIKey({
		apiKey: config.get('apiKey')
	});

    provider.setBackEndAddr(config.get('backEndAddress') || 'http://127.0.0.1:10024/api/search')

	provider.setSettings({
		pasteOnClick: config.get('pasteOnClick') || false,
		maxTokens: config.get('maxTokens') || 500,
		temperature: config.get('temperature') || 0.5,
		model: config.get('model') || 'text-davinci-003',
        enableTranslate: config.get('enableTranslate') || false,
        vendor: config.get('vendor') || 'openai',
        topP: config.get('topP') || 0.5
	});

	// 注册页面视图
	context.subscriptions.push(
		vscode.window.registerWebviewViewProvider(waltsViewProvider.viewType, provider,  {
			webviewOptions: { retainContextWhenHidden: true }
		})
	);


	const commandHandler = (command:string) => {
		provider.search('', command);
	};

	// 注册命令
	context.subscriptions.push(
		vscode.commands.registerCommand('walts.ask', () => 
			vscode.window.showInputBox({ prompt: 'Please ask me any question, and it is my pleasure to help you' })
			.then((value) => provider.search(value))
		),
		vscode.commands.registerCommand('walts.explain', () => commandHandler('code explain')),
		vscode.commands.registerCommand('walts.refactor', () => commandHandler('code refactor')),
		vscode.commands.registerCommand('walts.optimize', () => commandHandler('code optimize')),
		vscode.commands.registerCommand('walts.findProblems', () => commandHandler('code inspect')),
		vscode.commands.registerCommand('walts.documentation', () => commandHandler('code documentation'))
	);


	// 用户设置发生改变时保存新的用户设置
	vscode.workspace.onDidChangeConfiguration((event: vscode.ConfigurationChangeEvent) => {
		if (event.affectsConfiguration('walts.apiKey')) {
			const config = vscode.workspace.getConfiguration('walts');
			provider.setAPIKey({ apiKey: config.get('apiKey') });
			console.log("API key changed");
		} else if (event.affectsConfiguration('walts.pasteOnClick')) {
			const config = vscode.workspace.getConfiguration('walts');
			provider.setSettings({ pasteOnClick: config.get('pasteOnClick') || false });
		} else if (event.affectsConfiguration('walts.maxTokens')) {
			const config = vscode.workspace.getConfiguration('walts');
			provider.setSettings({ maxTokens: config.get('maxTokens') || 500 });
		} else if (event.affectsConfiguration('walts.temperature')) {
			const config = vscode.workspace.getConfiguration('walts');
			provider.setSettings({ temperature: config.get('temperature') || 0.5 });
		} else if (event.affectsConfiguration('walts.model')) {
			const config = vscode.workspace.getConfiguration('walts');
			provider.setSettings({ model: config.get('model') || 'text-davinci-003' });
		} else if (event.affectsConfiguration('walts.enableTranslate')) {
			const config = vscode.workspace.getConfiguration('walts');
			provider.setSettings({ enableTranslate: config.get('enableTranslate') || false });
		} else if (event.affectsConfiguration('walts.vendor')) {
			const config = vscode.workspace.getConfiguration('walts');
			provider.setSettings({ vendor: config.get('vendor') || 'openai' });
		} else if (event.affectsConfiguration('walts.topP')) {
			const config = vscode.workspace.getConfiguration('walts');
			provider.setSettings({ topP: config.get('topP') || 0.5 });
		} 
	});
}





class waltsViewProvider implements vscode.WebviewViewProvider {
	public static readonly viewType = 'walts.chatView';
	private _view?: vscode.WebviewView;

    private _historySessions?: string[] = []; 

	private _response?: string;
	private _prompt?: string;
	private _fullPrompt?: string;
	private _currentMessageNumber = 0;

    private _backEndAddr: string = '';

	private _settings: Settings = {
		pasteOnClick: true,
		maxTokens: 500,
		temperature: 0.5,
        enableTranslate: false
	};
	private _apiKey?: string;


	// 构造函数
	constructor(private readonly _extensionUri: vscode.Uri) {
        
	}

	// 设置APIKey
	public setAPIKey(apiKey: KeyInfo) {
		this._apiKey = apiKey.apiKey;
	}


    public setBackEndAddr(addr: string) {
        this._backEndAddr = addr
    }

	public setSettings(settings: Settings) {
		this._settings = {...this._settings, ...settings};
	}

	public getSettings() {
		return this._settings;
	}

	public resolveWebviewView(
		webviewView: vscode.WebviewView,
		context: vscode.WebviewViewResolveContext,
		_token: vscode.CancellationToken,
	) {
		this._view = webviewView;

		// 设置选项，允许JS脚本
		webviewView.webview.options = {
			enableScripts: true,
			localResourceRoots: [
				this._extensionUri
			]
		};

		// 设置html页面视图
		webviewView.webview.html = this._getHtmlForWebview(webviewView.webview);

		// 添加一个时间监听器，当页面收到消息时触发
		webviewView.webview.onDidReceiveMessage(data => {
			switch (data.type) {
				case 'codeSelected':
					{
						// 如果pasteOnClick设置被禁用不做任何事
						if (!this._settings.pasteOnClick) {
							break;
						}
						let code = data.value;
						//code = code.replace(/([^\\])(\$)([^{0-9])/g, "$1\\$$$3");
						const snippet = new vscode.SnippetString();
						snippet.appendText(code);
						// insert the code as a snippet into the active text editor
						vscode.window.activeTextEditor?.insertSnippet(snippet);
						break;
					}
				case 'prompt':
					{
						this.search(data.value, 'plain');
					}
			}
		});
  
	}


	public async resetSession() {
		this._prompt = '';
		this._response = '';
		this._fullPrompt = '';
		this._view?.webview.postMessage({ type: 'setPrompt', value: '' });
		this._view?.webview.postMessage({ type: 'addResponse', value: '' });
	}


	public async search(prompt?:string, promptType?:string) {
        const config = vscode.workspace.getConfiguration('walts');
		this._prompt = prompt;
		if (!prompt) {
			return;
		};


		// focus gpt activity from activity bar
		if (!this._view) {
			await vscode.commands.executeCommand('walts.chatView.focus');
		} else {
			this._view?.show?.(true);
		}
		
		let response = '';
		this._response = '';
		// Get the selected text of the active editor
		const selection = vscode.window.activeTextEditor?.selection;
		const selectedText = vscode.window.activeTextEditor?.document.getText(selection);

		// 确保提示已显示
		this._view?.webview.postMessage({ type: 'setPrompt', value: this._prompt });
		this._view?.webview.postMessage({ type: 'addResponse', value: '...' });


        // 请求后端获取回复
        
		// Increment the message number
		this._currentMessageNumber++;

		try {
            const res = await axios.post(this._backEndAddr, {
                "prompt": prompt,
                "model": this._settings.model,
                "vendor": this._settings.vendor,
                "APIKey": this._apiKey,
                "promptType": promptType,
                "code": selection,
                "maxTokens": this._settings.maxTokens,
                "temprature": this._settings.temperature,
                "enableTranslate": this._settings.enableTranslate,
                "topP": this._settings.topP
            })
            response = res.data
			// close unclosed codeblocks
			// Use a regular expression to find all occurrences of the substring in the string
			const REGEX_CODEBLOCK = new RegExp('\`\`\`', 'g');
			const matches = response.match(REGEX_CODEBLOCK);
			// Return the number of occurrences of the substring in the response, check if even
			const count = matches ? matches.length : 0;
			if (count % 2 !== 0) {
				//  append ``` to the end to make the last code block complete
				response += '\n\`\`\`';
			}
			response += `\n\n---\n`;

		} catch (error:any) {
			let e = '';
			if (error.response) {
				console.log(error.response.status);
				console.log(error.response.data);
				e = `${error.response.status} ${error.response.data.message}`;
			} else {
				console.log(error.message);
				e = error.message;
			}
			response += `\n\n---\n[ERROR] ${e}`;
		}
		

		// Saves the response
		this._response = response;
        console.log(response)
		// Show the view and send a message to the webview with the response
		if (this._view) {
			this._view.show?.(true);
			this._view.webview.postMessage({ type: 'addResponse', value: response });
		}
	}

	private _getHtmlForWebview(webview: vscode.Webview) {

		const scriptUri = webview.asWebviewUri(vscode.Uri.joinPath(this._extensionUri, 'media', 'main.js'));
		const microlightUri = webview.asWebviewUri(vscode.Uri.joinPath(this._extensionUri, 'media', 'scripts', 'microlight.min.js'));
		const tailwindUri = webview.asWebviewUri(vscode.Uri.joinPath(this._extensionUri, 'media', 'scripts', 'showdown.min.js'));
		const showdownUri = webview.asWebviewUri(vscode.Uri.joinPath(this._extensionUri, 'media', 'scripts', 'tailwind.min.js'));
		return `<!DOCTYPE html>
			<html lang="en">
			<head>
				<meta charset="UTF-8">
				<meta name="viewport" content="width=device-width, initial-scale=1.0">
				<script src="${tailwindUri}"></script>
				<script src="${showdownUri}"></script>
				<script src="${microlightUri}"></script>
				<style>
				.code {
					white-space: pre;
				}
				p {
					padding-top: 0.4rem;
					padding-bottom: 0.4rem;
				}
				/* overrides vscodes style reset, displays as if inside web browser */
				ul, ol {
					list-style: initial !important;
					margin-left: 10px !important;
				}
				h1, h2, h3, h4, h5, h6 {
					font-weight: bold !important;
				}
				</style>
			</head>
			<body>
				<input class="h-10 w-full text-white bg-stone-700 p-4 text-sm" placeholder="Ask GPT3 something" id="prompt-input" />
				<div id="response" class="pt-4 text-sm">
				</div>
				<script src="${scriptUri}"></script>
			</body>
			</html>`;
	}
}

// This method is called when your extension is deactivated
export function deactivate() {}

src/prompt.py
import json

import openai
import zhipuai
import SparkApi
from flask import Flask, request
from websocket import create_connection

from util import translate


app = Flask(__name__)
with open('config/config-dev.json', encoding='utf-8') as f:
    config = json.load(f)
appConfig = config.get('app')
promptConfig = config.get('prompt')
chinesePromptConfig = config.get('chinesePrompt')
apiKeyConfig = config.get('apiKey')
sparkAIConfig = config.get('sparkAI')
youdaoConfig = config.get('youdao')
openai.api_key = apiKeyConfig['openai']
zhipuai.api_key = apiKeyConfig['zhipuai']


@app.route('/api/search', methods=['POST'])
def search():
    # 接口数据获取
    request_data = request.get_json()
    vendor = request_data.get('vendor', 'openai')
    model = request_data.get('model', 'text-davinci-003')
    api_key = request_data.get('APIKey')
    prompt_type = request_data.get('promptType', 'plain')
    prompt = request_data.get('prompt')
    code = request_data.get('code')
    max_tokens = request_data.get('maxTokens', 1024)
    temperature = request_data.get('temprature', 0.5)
    enable_translate = request_data.get('enableTranslate', False)
    top_p = request_data.get('topP', 0.5)
    # 数据有效性检查
    assert prompt_type in ['plain', 'code explain', 'code refactor', 'code optimize', 'code inspect', 'code documentation']
    if vendor == 'openai':
        return searchOpenAI(model, api_key, prompt_type, prompt, code, max_tokens, temperature, enable_translate, top_p)
    elif vendor == 'zhipuai':
        return searchZhiPuAI(model, api_key, prompt_type, prompt, code, max_tokens, temperature, top_p)
    elif vendor == 'sparkai':
        return searchSparkAI(prompt_type, prompt, code)
    
    
def searchOpenAI(model, api_key, prompt_type, prompt, code, max_tokens, temperature, enable_translate, top_p):
    if api_key:
        openai.api_key = api_key
    assert model in ['text-davinci-003', 'text-davinci-002', 'text-curie-001', 'gpt-3.5-turbo', 'code-davinci-002', 'code-cushman-002']
    if enable_translate and prompt_type == 'plain':
        prompt = translate(prompt, youdaoConfig)
    prompt = buildPrompt(prompt_type, prompt, code, True)
    response = openai.Completion.create(
        model=model,
        prompt=prompt,
        max_tokens=max_tokens,
        temperature=temperature,
        top_p=top_p
    )
    openai.api_key = apiKeyConfig['openai']
    return response.choices[0].text


def searchZhiPuAI(model, api_key, prompt_type, prompt, code, max_tokens, temperature, top_p):
    model = "chatglm_turbo"
    if api_key:
        zhipuai.api_key = api_key
    prompt = buildPrompt(prompt_type, prompt, code, False)
    response = zhipuai.model_api.invoke(
        model=model,
        prompt=[{"role": "user", "content": prompt}],
        top_p=top_p,
        temperature=temperature,
        max_tokens=max_tokens
    )
    zhipuai.api_key = apiKeyConfig['zhipuai']
    return response['data']['choices'][0]['content']


def searchSparkAI(prompt_type, prompt, code):
    SparkApi.answer = ""
    prompt = buildPrompt(prompt_type, prompt, code, False)
    SparkApi.main(sparkAIConfig['appid'], sparkAIConfig['api_key'], sparkAIConfig['app_secret'], sparkAIConfig['url'], sparkAIConfig['domain'], [{"role": "user", "content": prompt}])
    return SparkApi.answer
    

def buildPrompt(prompt_type, prompt, code, english):
    if english:
        if prompt_type == 'code explain':
            prompt = promptConfig['codeExplainPrompt']
        elif prompt_type == 'code refactor':
            prompt = promptConfig['codeRefactorPrompt']
        elif prompt_type == 'code optimize':
            prompt = promptConfig['codeOptimizePrompt']
        elif prompt_type == 'code inspect':
            prompt = promptConfig['codeInspectPrompt']
        elif prompt_type == 'code documentation':
            prompt = promptConfig['codeDocumentationPrompt']
        if code:
            prompt = f"${prompt}\n\`\`\`\n${code}\n\`\`\`"
        prompt = promptConfig['promptTemplate'].format(prompt)
    else:
        if prompt_type == 'code explain':
            prompt = chinesePromptConfig['codeExplainPrompt']
        elif prompt_type == 'code refactor':
            prompt = chinesePromptConfig['codeRefactorPrompt']
        elif prompt_type == 'code optimize':
            prompt = chinesePromptConfig['codeOptimizePrompt']
        elif prompt_type == 'code inspect':
            prompt = chinesePromptConfig['codeInspectPrompt']
        elif prompt_type == 'code documentation':
            prompt = chinesePromptConfig['codeDocumentationPrompt']
        if code:
            prompt = f"${prompt}\n\`\`\`\n${code}\n\`\`\`"
        prompt = chinesePromptConfig['promptTemplate'].format(prompt)
    return prompt
    

if __name__ == '__main__':
    app.run(host=appConfig['host'], port=appConfig['port'])
    #app.run(host='10.119.6.207', port=1024)